HTTP/2 edits (#1618)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/http2.md

> reviewers: [LPardue, bazzadp, ibnesayeed]

Author: The [doc](https://docs.google.com/document/d/1ybIBEtM-C2e-pEZMs_bp39mtEs3u6s8LHOz6Es8QUGA/edit#heading=h.xleqpbggrlck) lists @LPardue as an author, not reviewer.

Which is correct? @dotjs @LPardue

Reviewer: I wanted to author but the other folks did such a good job all I could do is review 😀

> analysts: [gregorywolf]

Author: @gregorywolf is also listed as a reviewer in the [doc](https://docs.google.com/document/d/1ybIBEtM-C2e-pEZMs_bp39mtEs3u6s8LHOz6Es8QUGA/edit#heading=h.xleqpbggrlck). Is that correct? @dotjs

Reviewer 2: Didn't get time to review and pinged me the other day to ask to be credited just as an analyst because of that. I told him @dotjs had already ruthlessly cut him from the reviewer credit 😉

translators: []
dotjs_bio: Andrew works at <a href="https://www.cloudflare.com/">Cloudflare</a> helping to make the web faster and more secure. He spends his time deploying, measuring and improving new protocols and asset delivery to improve end-user web site performance.
rmarx_bio: Robin is a web protocol and performance researcher at <a href="https://www.uhasselt.be/edm">Hasselt University, Belgium</a>. He has been working on getting QUIC and HTTP/3 ready to use by creating tools like <a href="https://github.com/quiclog">qlog and qvis</a>.

> MikeBishop_bio: Editor of HTTP/3 with @quicwg. Architect in Akamai's Foundry group.

Author: @MikeBishop I grabbed this from your GitHub profile, but please suggest a different bio if you'd prefer.

Reviewer 2: In case @MikeBishop doesn't get back let's change the QUIC WG link since it's no longer on GitHub:

```suggestion
MikeBishop_bio: Editor of HTTP/3 with the [QUIC Working Group](https://quicwg.org/). Architect in [Akamai](https://www.akamai.com/)'s Foundry group.
```

      <tr>
        <th>Protocol</th>
        <th>Desktop</th>
        <th>Mobile</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>HTTP/1.0</td>
        <td class="numeric">0.06%</td>
        <td class="numeric">0.05%</td>
      </tr>
      <tr>
        <td>HTTP/1.1</td>
        <td class="numeric">49.22%</td>
        <td class="numeric">50.05%</td>
      </tr>
      <tr>
        <td>HTTP/2</td>
        <td class="numeric">49.97%</td>
        <td class="numeric">49.28%</td>
      </tr>
    </tbody>
  </table>

  <figcaption>{{ figure_link(caption="HTTP version usage for home pages.", sheets_gid="1447413141", sql_file="measure_of_all_http_versions_for_main_page_of_all_sites.sql") }}</figcaption>
</figure>

gQUIC is not seen in the home page data for two reasons. To measure a website over gQUIC, the HTTP Archive crawl would have to perform protocol negotiation via the [alternative services](#alternative-services) header and then use this endpoint to load the site over gQUIC. This was not supported this year, but expect it to be available in next year's Web Almanac. Also, gQUIC is predominantly used for third-party Google tools rather than serving home pages.

> {# TODO(authors, analysts): Where does the 150% stat come from? Can we link to it or show it as a figure? #}

Author: @bazzadp @dotjs

Reviewer 2: Is that from here @dotjs : https://httparchive.org/reports/state-of-the-web#pctHttps. Agree it should be linked.

The drive to increase [security](./security) and [privacy](./privacy) on the web has seen requests over TLS increase by over 150% in the last 4 years. Today, over 86% of all requests on mobile and desktop are encrypted. Looking only at home pages, the numbers are still an impressive 78.1% of desktop and 74.7% of mobile. This is important because HTTP/2 is only supported by browsers over TLS. The proportion served over HTTP/2, as shown in Figure 22.5, has also increased by 10 percentage points from [last year](../2019/http2#fig-5), from 55% to 65%.

<figure>
  <table>
    <thead>
      <tr>
        <th>Protocol</th>
        <th>Desktop</th>
        <th>Mobile</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>HTTP/1.1</td>
        <td class="numeric">36.05%</td>
        <td class="numeric">34.04%</td>
      </tr>
      <tr>
        <td>HTTP/2</td>
        <td class="numeric">63.95%</td>
        <td class="numeric">65.96%</td>
      </tr>
    </tbody>
  </table>

>   {# TODO(authors): Last year's version of this stat (linked above) used "HTTPS home pages" instead of "TLS home pages". Are they interchangeable? Is it better to be consistent or is TLS more correct? #}

Author: @dotjs

Reviewer 2: This is one which @rmarx and I have argued with in the past 😁

I prefer HTTPS unless talking specifically about the TLS protocol features (ciphers, ALPN...etc.). I feel it's friendlier, more atune to what the reader sees in the URL bar and avoids the whole SSL/TLS noise and mouthful (SSL is now inaccurate but is more widely known than TLS to non-techies, SSL/TLS is too much of a mouthful).

Also what we use in HTTP Archive: https://httparchive.org/reports/state-of-the-web#pctHttps

Reviewer 3: Yeah, I agree with @bazzadp on this one. While TLS is more "correct", HTTPS covers the load well enough here and is more approachable for the wide audience the Almanac is targeting. 

Though I still feel a book on HTTP/2 should use TLS instead, Barry ;)

  description="A bar chart comparing the fraction of third-party HTTP/2 or gQUIC requests by category. All third-party requests for all websites are served over HTTP/2 or gQUIC for 95% of tag managers, 90% of analytics and CDN, 80% of ads, social, hosting and utility, 40% of marketing and 30% of customer-success.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vSOkWXtrbMfTLdhlKbBGDjRU3zKbnCQi3iPhfuKaFs5mj4smEzInDCYEnk63gBdgsJ3GFk2gf4FOKCU/pubchart?oid=1419102835&format=interactive",
  sheets_gid="1059610651",
  sql_file="http2_3rd_party_by_types.sql"
)
}}

### Server support

Browser auto-update mechanisms are a driving factor for client-side adoption of new web standards. It's [estimated](https://caniuse.com/http2) that over 97% of global users support HTTP/2, up slightly from 95% measured last year.

Unfortunately, the upgrade path for servers is more difficult, especially with the requirement to support TLS. For mobile and desktop, we can see from Figure 22.10, that he majority of HTTP/2 sites are served by nginx, Cloudflare, and Apache. Almost half of the HTTP/1.1 sites are served by Apache.

> 
> {# TODO(analysts, authors): Why is this figure in terms of the number of websites, rather than percent? Should the server values also be capitalized accordingly? (eg Cloudflare rather than cloudflare) #}

Author: ```suggestion

```

  sql_file="http2_%_1st_party_vs_3rd_party.sql",
  sheets_gid="733872185"
)
}}

Figure 22.8 shows the distribution of known third-party HTTP/2 or gQUIC requests by content-type. For example 90% of websites serve 100% of third party fonts and audio over HTTP/2 or gQUIC, only 5% over HTTP/1 and 5% are a mix. The majority of third-party assets are either scripts or images, and are solely served over HTTP/2 or gQUIC on 60% and 70% of websites respectively.

{{ figure_markup(
  image="http2-third-party-http2-usage-by-content-type.png",
  caption="The fraction of known third-party HTTP/2 or gQUIC requests by content-type per website.",
  description="A bar chart comparing the fraction of third-party HTTP/2 requests by content-type. All third-party requests are served over HTTP/2 or gQUIC for 90% of audio and fonts, 80% of css and video, 70% of html, image and text and 60% of scripts.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vSOkWXtrbMfTLdhlKbBGDjRU3zKbnCQi3iPhfuKaFs5mj4smEzInDCYEnk63gBdgsJ3GFk2gf4FOKCU/pubchart?oid=1264128523&format=interactive",
  sheets_gid="419557288",
  sql_file="http2_%_1st_party_vs_3rd_party_by_type.sql"
)
}}

> {# TODO(authors, analysts): Is this breakdown specifically looking at mobile? I don't see a notice that all data is mobile unless otherwise noted, so it's worth clarifying here. #}

Author: @dotjs @bazzadp

Reviewer 2: Look to be mobile. We should add a note in the intro to say, "where results are similar between desktop and mobile, stats are presented from the mobile dataset." or something like that.

Reviewer 4: Yes As @bazzadp told me it's a mobile first web. I'll add a note in the intro.

  caption="The fraction of known third-party HTTP/2 or gQUIC requests by category per website.",
  description="A bar chart comparing the fraction of third-party HTTP/2 or gQUIC requests by category. All third-party requests for all websites are served over HTTP/2 or gQUIC for 95% of tag managers, 90% of analytics and CDN, 80% of ads, social, hosting and utility, 40% of marketing and 30% of customer-success.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vSOkWXtrbMfTLdhlKbBGDjRU3zKbnCQi3iPhfuKaFs5mj4smEzInDCYEnk63gBdgsJ3GFk2gf4FOKCU/pubchart?oid=1419102835&format=interactive",
  sheets_gid="1059610651",
  sql_file="http2_3rd_party_by_types.sql"
)
}}

### Server support

Browser auto-update mechanisms are a driving factor for client-side adoption of new web standards. It's [estimated](https://caniuse.com/http2) that over 97% of global users support HTTP/2, up slightly from 95% measured last year.

Unfortunately, the upgrade path for servers is more difficult, especially with the requirement to support TLS. For mobile and desktop, Figure 22.10 shows the breakdown of HTTP server headers returned for sites that support HTTP/2 and HTTP/1.

> {# TODO(analysts, authors): Why is this figure in terms of the number of websites, rather than percent? Should the server values also be capitalized accordingly? (eg Cloudflare rather than cloudflare) #}

Author: @dotjs @bazzadp

Reviewer 2: This was @dotjs but I actually think I agree with him here.

I think a percentage in terms of all sites, or even relative to that particular server, would be misleading and lose the fact that nginx and and apache have way more sites than gse for example.

As to the capitalisation these are the server headers (as stated in the line above the graph) rather than company names and looks like this year we normalised to avoid any case issues. I think it's fine as is.

Reviewer 4: This was discussed with the authors/analysts. Using the absolute number of websites allows the reader to see the proportion of h2 vs h1 for each server and also see the relative proportion of websites served per sever.

The majority of HTTP/2 sites are served by nginx, Cloudflare, and Apache. Almost half of the HTTP/1.1 sites are served by Apache.

How has HTTP/2 adoption changed in the last year for each server? Figure 22.11 shows the percentage of server installs that use HTTP/2 in 2019 and 2020. We see a general adoption increase of around 10% across all servers. Apache and IIS are still under 25% HTTP/2. This suggests that either new servers tend to be nginx or it is seen as too difficult or not worthwhile to upgrade Apache or IIS to HTTP/2 and/or TLS.

> {# TODO(analysts, authors): Ditto on server capitalization. #}

Author: @dotjs @bazzadp

Reviewer 2: Ditto right back at you 😉

  sheets_gid="306338094",
  sql_file="count_of_h2_and_h3_sites_grouped_by_server.sql"
  )
}}

A long-term recommendation to improve website performance has been to use a CDN. The benefit is a reduction in latency by both serving content and terminating connections closer to the end user. This helps mitigate the rapid evolution in protocol deployment and the additional complexities in tuning servers and operating systems (see the [Prioritization](#prioritization) section for more details). To utilize the new protocols effectively, using a CDN can be seen as the recommended approach.

CDNs can be classed in two broad categories: those that serve the home page and/or asset subdomains, and those that are mainly used to serve third-party content. Examples of the first category are the larger generic CDNs (such as Cloudflare, Akamai, or Fastly) and the more specific (such as WordPress or Netlify). Looking at the difference in HTTP/2 adoption rates for home pages served with or without a CDN, we see:

- **80%** of mobile home pages are served over HTTP/2 if a CDN is used
- **30%** of mobile home pages are served over HTTP/2 if a CDN is not used

> {# TODO(authors, editors): Throughout the chapter there are several lead-ins to figures that are very similar/redundant with the existing figure captions. Is it worth writing these out? #}

Author: @dotjs

Reviewer 4: Agreed they were written before adding the captions. I will review.

{{ figure_markup(
  image="http2-cdn-http2-usage.png",
  caption="Comparsion of HTTP/2 and gQUIC usage for websites using a CDN.",
  description="A line chart comparing the fraction of reqests served using HTTP/2 or gQUIC for websites that use a CDN compared to sites that do not. The x-axis show the percentiles of web page ordered by percentage of requests. 23% of websites that do not use a CDN have no HTTP/2 or gQUIC usage. In comparision the 60% of websites using a CDN have all HTTP/2 or gQUIC usage. 93% of websites that use a CDN and 47% of non-CDN sites have 50% or more HTTP/2 or gQUIC usage.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vSOkWXtrbMfTLdhlKbBGDjRU3zKbnCQi3iPhfuKaFs5mj4smEzInDCYEnk63gBdgsJ3GFk2gf4FOKCU/pubchart?oid=1779365083&format=interactive",
  sheets_gid="1457263817",
  sql_file="cdn_summary.sql"
)
}}

Figure 22.13 shows the difference in HTTP/2 and gQUIC adoption when a website is using a CDN. 70% of pages use HTTP/2 for all third-party requests when a CDN is used. Without a CDN, only 25% of pages use HTTP/2 for all third-party requests.

> {# TODO(authors): What do you hope readers get out of this figure? There's no discussion/interpretation of the results. #}

Author: @dotjs

<figcaption>{{ figure_link(caption="Percentage of HTTP/2 requests served by the third-party resouce CDNs over mobile", sheets_gid="781660433", sql_file="cdn_detail_by_cdn.sql") }}</figcaption>
</figure>

> {# TODO(authors, editors): Sort of a nit. Headings inconsistently ask questions and give short titles to describe the contents. It would be more natural to have a unified naming convention for headers. #}

Author: @dotjs

Reviewer 4: Updated.

## How is HTTP/2 performing?

Measuring the impact of how a protocol is performing is difficult with the current HTTP Archive [approach](./methodology). It would be really fascinating to be able to quantify the impact of concurrent connections, the effect of packet loss, and differrent congestion control mechanisms. To really compare performance, each website would have to be crawled over each protocol over different network conditions. What we can do instead is to look into the impact on the number of connections a website uses.

### Reducing connections

As discussed [earlier](#http10-to-http2), HTTP/1.1 only allows a single request at a time over a TCP connection. Most browsers get around this by allowing six parallel connections per host. The major improvement with HTTP/2 is that multiple requests can be multiplexed over a single TCP connection.  This should reduce the total number of connections—and the associated time and resources—required to load a page.

> {# TODO(analysts): Nit, but I think it'd make more sense to show 2020 to the right of 2016 as I think most readers will conceptualize time going from left to right. #}

Author: @bazzadp

Reviewer 2: Fair point. Fixed the graphs affected and added fall back images in https://github.com/HTTPArchive/almanac.httparchive.org/pull/1618/commits/d825f2b5a14babe878e1f3a087cc7ecce159994d

)
}}

> {# TODO(authors, analysts): Clarify if this data is mobile-only. Also clarify what metric "the number of requests has only dropped from 74 to 73" is referring to. Median requests per page? #}

Author: @dotjs @bazzadp

        <td class="numeric">36.34</td>
        <td class="numeric">1</td>
        <td class="numeric">0.06</td>
      </tr>
      <tr>
        <td class="numeric">50</td>
        <td class="numeric">3</td>
        <td class="numeric">183.83</td>
        <td class="numeric">2</td>
        <td class="numeric">24.06</td>
      </tr>
      <tr>
        <td class="numeric">75</td>
        <td class="numeric">10</td>
        <td class="numeric">225.41</td>
        <td class="numeric">5</td>
        <td class="numeric">204.65</td>
      </tr>
      <tr>
        <td class="numeric">90</td>
        <td class="numeric">12</td>
        <td class="numeric">351.05</td>
        <td class="numeric">18</td>
        <td class="numeric">453.57</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Distribtion of pushed assets on mobile.", sheets_gid="698874709", sql_file="number_of_h2_and_h3_pushed_resources_and_avg_bytes.sql") }}</figcaption>
</figure>

Looking at the frequency of push by content type in Figure 22.20, we see 89% of pages push scripts and 68% push CSS. This makes sense, as these can be small files typically on the critical path to render a page.

> {# TODO(analysts): Should this chart differentiate desktop and mobile? #}

Author: @bazzadp

Reviewer 2: Think it should but struggling to see how @dotjs created this chart to be honest so will let him decide and then fix and I can retake backup image then.

Reviewer 4: Added desktop @bazzadp Could you update the image please.

Reviewer 2: Done.

HTTP/3 and QUIC are highly advanced protocols with powerful performance and security features, such as a new HTTP prioritization system, HOL blocking removal, and 0-RTT connection establishment. This sophistication also makes them difficult to deploy and use correctly, as has turned out to be the case for HTTP/2. We predict that early deployments will mainly be done via the early adoption of CDNs such as Cloudflare, Fastly and Akamai. This will probably mean that a large part of HTTP/2 traffic can and will be upgraded to HTTP/3 automatically in 2021, giving the new protocol a large traffic share almost out of the box. When and if smaller or individual deployments will follow suit is yet more difficult to answer. Most likely, we will continue to see a healthy mix of HTTP/3, HTTP/2 and even HTTP/1.1 on the web for years to come.

## Conclusion

> This year HTTP/2 has become the dominant protocol serving 64% of all requests, having grown by 10% during the year. Home pages have seen a 13% increase in HTTP/2 adoption, leading to an even split of pages served over HTTP/1 and HTTP/2. Using a CDN to serve your home page pushes HTTP/2 adoption up to 80% compared with 30% for non CDN usage. There remain some older servers: apache and IIS, that are proving resistant to upgrading to HTTP/2 and TLS. A big success has been the decrease in website connection usage due to HTTP/2 connection multiplexing. The median number of connections in 2016 was 23 compared to 13 in 2020.

Reviewer: ```suggestion
This year HTTP/2 has become the dominant protocol serving 64% of all requests, having grown by 10% during the year. Home pages have seen a 13% increase in HTTP/2 adoption, leading to an even split of pages served over HTTP/1 and HTTP/2. Using a CDN to serve your home page pushes HTTP/2 adoption up to 80% compared with 30% for non-CDN usage. There remain some older servers: apache and IIS, that are proving resistant to upgrading to HTTP/2 and TLS. A big success has been the decrease in website connection usage due to HTTP/2 connection multiplexing. The median number of connections in 2016 was 23 compared to 13 in 2020.
```

      <tr>
        <td class="numeric">90</td>
        <td class="numeric">29</td>
        <td class="numeric">32</td>
        <td class="numeric">4</td>
        <td class="numeric">24</td>
        <td class="numeric">23</td>
        <td class="numeric">1</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Distribtion of number pushed asset by content type on mobile.", sheets_gid="238923402", sql_file="number_of_h2_and_h3_pushed_resources_and_bytes_by_content_type.sql") }}</figcaption>
</figure>

Pushing of video dominates the higher percentiles of bandwidth use. This can, however, negatively impact page performance by contending with the other resources required to render the page.

<figure>
  <table>
    <thead>
      <tr>
        <th>Percentile</th>
        <th>css</th>
        <th>font</th>
        <th>html</th>
        <th>image</th>
        <th>script</th>
        <th>video</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class="numeric">10</td>
        <td class="numeric">21.1</td>
        <td class="numeric">784.7</td>
        <td class="numeric">1.5</td>
        <td class="numeric">4.1</td>
        <td class="numeric">4.0</td>

>         {# TODO(analysts): Please double check the video stats, which is 3444.3 for all percentiles. #}

Author: @bazzadp

Reviewer 2: Looks to me like only one site pushes video. This means where video is pushed (that one site), all percentiles push the same amount.

Not sure how best to account for that. Any ideas @dotjs or @rviscomi ?

Reviewer 4: I had the same initial confusion. As mentioned if only a single website is pushing video then the value is the same across all percentiles. I could add a sentence to explain ?

Reviewer 4: Removed section. On reading again I don't think it adds much.

)
}}

It's important to note that this low adoption only reflects gQUIC and HTTP/3 usage for loading Web pages. For several years already, Facebook has had a much more extensive deployment of IETF QUIC and HTTP/3 for loading data inside of its native applications. [QUIC and HTTP/3 already make up over 75% of their total internet traffic](https://engineering.fb.com/2020/10/21/networking-traffic/how-facebook-is-bringing-quic-to-billions/). It is clear that HTTP/3 is only just getting started!

Now you might wonder: if not everyone is already using HTTP/2, why would we need HTTP/3 so soon? What benefits can we expect in practice? Let's take a closer look at its internal mechanisms.

> {# TODO(authors): I honestly found this section really interesting, but it doesn't contain any real-world data. How can we add an element of data to this? #}

Author: @dotjs

Finally, TLS is already used to protect HTTP/1.1 and HTTP/2. QUIC, however, has a deep integration of TLS 1.3, protecting both HTTP/3 data and QUIC packet metadata, such as packet numbers. Using TLS in this way improves end-user privacy and security and makes continued protocol evolution easier. Combining the transport and cryptographic handshakes means that connection setup takes just a single RTT, compared to TCP's minimum of two and worst case of four. In some cases, QUIC can even go one step further and send HTTP data along with its very first message, which is called _0-RTT_. These fast connection setup times are expected to really help HTTP/3 outperform HTTP/2.

#### So, will HTTP/3 help?

> {# TODO(authors): Who is "we" at the end of this paragraph? Worth clarifying for the readers, or making it more ambiguous like "the web community" or "implementers". #}

Author: @dotjs

On the surface, HTTP/3 is really not all that different from HTTP/2. It doesn't add any major features, but mainly changes how the existing ones work under the surface. The real improvements come from QUIC, which offers faster connection setups, increased robustness, and resilience to packet loss. As such, HTTP/3 is expected to do better than HTTP/2 on worse networks, while offering very similar performance on faster systems. However, that is if we can get HTTP/3 working, which can be challenging in practice.

### Deploying and discovering QUIC and HTTP/3

> {# TODO(authors): Where exactly is "HTTP Alternative Services" described earlier? This is the first instance of it in the chapter. #}

Author: @dotjs

Reviewer 4: There is a section (now linked)

There is ongoing work to define ways to discover HTTP/3 without needing the TCP step. This should be considered an optimization though, as the UDP blocking issues are likely to mean that TCP-based HTTP sticks around. The [HTTPS DNS record](https://tools.ietf.org/html/draft-ietf-dnsop-svcb-https), is similar to HTTP Alternative Services and some CDNs are already [experimenting with these records](https://blog.cloudflare.com/speeding-up-https-and-http-3-negotiation-with-dns/).  In the long run, when most servers offer HTTP/3, browsers might switch to attempting that by default. But that will take a long time.

> {# TODO(authors, analysts): Should 41% or 60% be figures here? Are they HTTP Archive stats? There has been a bit of a drought of data/figures towards the end, so it may be good to end with one. #}

Author: @dotjs @bazzadp

Reviewer 2: As per https://github.com/HTTPArchive/almanac.httparchive.org/pull/1618#discussion_r534831196 I think these should be linked.

Reviewer 4: figure created

QUIC is dependent on TLS1.3, which is used for around 41% of requests, which means the other 60% of requests will need to update their TLS stack to support HTTP/3.

#### Dual stack deployment implications

> {# TODO(authors): This section is an awkward fit because it is a meta discussion about our inability to test, and it doesn't introduce any new data. Perhaps consider dropping? #}

Author: @dotjs

Reviewer 4: removed
